source/model/backtraceMetrics.ts

```diff
@@ -0,0 +1,228 @@
+import { BacktraceClientOptions } from '..';
+import { currentTimestamp, getBacktraceGUID, getEndpointParams, uuid } from '../utils';
+import { APP_NAME, USER_AGENT, VERSION } from '../consts/application';
+import { SEC_TO_MILLIS } from '../consts';
+
+/**
+ * Handles Backtrace Metrics.
+ */
+export class BacktraceMetrics {
+  private readonly universe: string;
+  private readonly token: string;
+  private readonly timeout: number  = 15000; // Fifteen seconds in milliseconds.
+  private readonly persistenceInterval: number  = 1800000; // Thirty minutes in milliseconds.
+  private readonly heartbeatInterval: number  = 60000; // One minutes in milliseconds.
+
+  private readonly timestamp = currentTimestamp()
+  private readonly userAgent = USER_AGENT;
+  private readonly applicationName = APP_NAME;
+  private readonly applicationVersion = VERSION;
+
+  private readonly guid = getBacktraceGUID();
+
+  public hostname = 'https://events.backtrace.io';
+
+  constructor(configuration: BacktraceClientOptions) {
+    if (!configuration.endpoint) {
+      throw new Error(`Backtrace: missing 'endpoint' option.`);
+    }
+    const {universe, token} = getEndpointParams(configuration.endpoint)
+
+    if(!universe) {
+      throw new Error(`Backtrace: 'universe' could not be parsed from the endpoint.`);
+    }
+
+    if(!token && !configuration.token) {
+      throw new Error(`Backtrace: missing 'token' option or it could not be parsed from the endpoint.`);
+    }
+
+    this.universe = universe;
+    this.token = configuration.token || token || ''; // default '' case needed since type checking isn't certain that type can not be undefined.
```

Comment:
we check if configuration token/token is defined in line 35. Do we need to assume token is '' ? I think we should drop this case.

Reply:
Typescript is not able to recognize that the condition `if !A && !B == false` guarantee that `A || B == true`, I think because it does the type checking per variable, and the conditional is compound so neither is guaranteed to be defined.

<img width="999" alt="Screen Shot 2022-01-07 at 1 16 42 PM" src="https://user-images.githubusercontent.com/47638521/148588526-c6b2e5f0-8bc3-41f2-a319-cf2b2cc26927.png">

If I refactor and make `if(!A && !B)` into two different ifs like `if(!A) {...}` and another `if(!B) {...}` then it won't complain anymore, but right now TS throws error unless I add default case `|| ''` 


If there is another solution?

Reply:
Yes - typescript, since we know that configuration.token & token exist, we can cast it to string

`(configuration.token || token) as string`